consensus/replay.go

```diff
@@ -276,11 +276,17 @@ func (h *Handshaker) ReplayBlocks(
 
 	// If appBlockHeight == 0 it means that we are at genesis and hence should send InitChain.
 	if appBlockHeight == 0 {
-		validators := make([]*types.Validator, len(h.genDoc.Validators))
-		for i, val := range h.genDoc.Validators {
-			validators[i] = types.NewValidator(val.PubKey, val.Power)
+		// Empty validator set is allowed in the genesis, because it can be set by the app in InitChain
+		var validatorSet *types.ValidatorSet
+		if len(h.genDoc.Validators) > 0 {
+			validators := make([]*types.Validator, len(h.genDoc.Validators))
+			for i, val := range h.genDoc.Validators {
+				validators[i] = types.NewValidator(val.PubKey, val.Power)
+			}
+			validatorSet = types.NewValidatorSet(validators)
+		} else {
+			validatorSet = types.NewValidatorSet(nil)
```

Comment:
Here in NewValidatorSet(main branch), nil and empty slice are treated separately.
Empty slice will cause panic.

func NewValidatorSet(valz []*Validator) *ValidatorSet {
	if valz != nil && len(valz) == 0 {
		panic("validator set initialization slice cannot be an empty slice (but it can be nil)")
	}

Reply:
That was changed (recently) on develop:
https://github.com/tendermint/tendermint/blob/0f96bea41dc54c99eab71b3f4bf7e11e3347fe48/types/validator_set.go#L45-L59

Reply:
I would suggest to remove the changes above (and only keep the changes below). But no biggie. Semantically, the code does the same and maybe it's more explicit with the `nil`.

Reply:
That's fine. Let me remove the first change as this PR will only go to `develop`.